### PR TITLE
Reorganise the Interpreter Root

### PR DESCRIPTION
**Pull request by iamrecursion**
### Pull Request Description
In order to properly support static analysis and compilation, the root of the interpreter had to be reorganised to have a distinct 'compiler' phase. This PR does so. 

### Important Notes
There are two kinds of `EnsoRootNode`, one which represents the root of the whole _program_, while the other represents executable roots such as closures (functions, lambdas, etc).

This PR also disables benchmarks on CI as there is no regression testing currently. They are still checked for if they compile to prevent regressions in that sense. 

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: Interpreter/src/main/java/org/enso/interpreter/node/ClosureRootNode.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import com.oracle.truffle.api.frame.FrameSlotKind;
+import com.oracle.truffle.api.frame.FrameUtil;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.source.SourceSection;
+import org.enso.interpreter.Language;
+import org.enso.interpreter.runtime.callable.function.Function;
+import org.enso.interpreter.runtime.state.Stateful;
+
+/**
+ * This node represents the root of Enso computations.
```
Comment (kustosz) [inline Interpreter/src/main/java/org/enso/interpreter/node/ClosureRootNode.java]: This is a very misleading and stale doc at this point

File: Interpreter/src/main/java/org/enso/interpreter/node/EnsoRootNode.java [hunk 2]
```diff
@@ -2,158 +2,100 @@
...
-    } else {
-      return result;
-    }
+  public String toString() {
+    return this.name;
   }
 
   /**
-   * Converts this node to a textual representation good for debugging.
+   * Gets a reference to the language with which this node is associated.
```
Comment (kustosz) [inline Interpreter/src/main/java/org/enso/interpreter/node/EnsoRootNode.java]: 1. It returns the language, not the language reference.
2. It should return a language reference, in which case it already exists, see https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/nodes/Node.html#lookupLanguageReference-java.lang.Class-
Shortly – the language should be accessed the same way context is.

File: Interpreter/src/main/java/org/enso/interpreter/node/EnsoRootNode.java [hunk 3]
```diff
@@ -2,158 +2,100 @@
...
    */
-  public void setTail(boolean isTail) {
-    body.setTail(isTail);
+  public FrameSlot getStateFrameSlot() {
+    return this.stateFrameSlot;
   }
 
   /**
-   * Returns the frame slot reference to state variable.
+   * Gets a reference to the source code represented by this node.
```
Comment (kustosz) [inline Interpreter/src/main/java/org/enso/interpreter/node/EnsoRootNode.java]: I'd rather avoid using `reference` in docs like this. Logically it returns the source section. Obviously there's no such thing as returning a value in java, everything is a reference, so it's not really worth mentioning. When I see `Gets a reference` I immediately think about some `SourceSectionReference` type that wraps a source section.

File: Interpreter/src/main/java/org/enso/interpreter/node/ProgramRootNode.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+  /**
+   * Executes the static analysis passes before executing the resultant program.
+   *
+   * @param frame the stack frame to execute in
+   * @return the result of executing this node
+   */
+  @Override
+  public Object execute(VirtualFrame frame) {
+    Context context = this.getLanguage().getCurrentContext();
```
Comment (kustosz) [inline Interpreter/src/main/java/org/enso/interpreter/node/ProgramRootNode.java]: Use  the `getContext()` method that is used on line 65 in the same method, or refactor both into a `@CachedContext` argument.

File: Interpreter/src/main/java/org/enso/interpreter/node/ProgramRootNode.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+    return this.ensoProgram.executeGeneric(frame);
+  }
+
+  /* Note [Static Passes]
+   * ~~~~~~~~~~~~~~~~~~~~
+   * Almost all of the static analysis functionality required by the interpreter requires access to
+   * the interpreter to execute small amounts of code. This is for purposes such as:
+   * - Type-level computation and evaluation during typechecking.
+   * - CTFE for optimisation.
```
Comment (kustosz) [inline Interpreter/src/main/java/org/enso/interpreter/node/ProgramRootNode.java]: I didn't immediately know what `CTFE` is and my first google search said `Chlorotrifluoroethylene`. Please expand this acronym the first time it's used in a document.

File: Interpreter/src/main/java/org/enso/interpreter/node/ProgramRootNode.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+  @Override
+  public Object execute(VirtualFrame frame) {
+    Context context = this.getLanguage().getCurrentContext();
+
+    // Note [Static Passes]
+    if (this.ensoProgram == null) {
+      this.ensoProgram = this.insert(context.compiler().run(this.sourceCode));
+    }
+
+    this.ensoProgram.setTail(programShouldBeTailRecursive);
```
Comment (kustosz) [inline Interpreter/src/main/java/org/enso/interpreter/node/ProgramRootNode.java]: This should be under that `this.ensoProgram == null` condition, as it should only be run once and is actually unsafe to run here – it modifies a `@CompilationFinal` field without a deopt.

File: Interpreter/src/main/java/org/enso/interpreter/node/ProgramRootNode.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+   * @param frame the stack frame to execute in
+   * @return the result of executing this node
+   */
+  @Override
+  public Object execute(VirtualFrame frame) {
+    Context context = this.getLanguage().getCurrentContext();
+
+    // Note [Static Passes]
+    if (this.ensoProgram == null) {
+      this.ensoProgram = this.insert(context.compiler().run(this.sourceCode));
```
Comment (kustosz) [inline Interpreter/src/main/java/org/enso/interpreter/node/ProgramRootNode.java]: Modifying a `@Child` needs a deopt.

File: Interpreter/src/main/java/org/enso/interpreter/node/callable/function/CreateFunctionNode.java [hunk 8]
```diff
@@ -35,7 +35,7 @@ public CreateFunctionNode(RootCallTarget callTarget, ArgumentDefinition[] args)
    */
   @Override
   public void setTail(boolean isTail) {
-    ((EnsoRootNode) callTarget.getRootNode()).setTail(isTail);
+    ((ClosureRootNode) callTarget.getRootNode()).setTail(isTail);
```
Comment (kustosz) [inline Interpreter/src/main/java/org/enso/interpreter/node/callable/function/CreateFunctionNode.java]: In general, the whole `setTail` stuff is not cool anymore. Every `setTail` needs a deopt. Before, we only used it _before_ execution (during node construction). Now that it's a part of `execute`, we need to safeguard them with deopts.

File: Syntax/definition/src/main/scala/org/enso/syntax/text/Debug.scala [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+              case ')' => go(i + 1, rest - 1, ss, bias + 1)
+              case _   => go(i + 1, rest - 1, ss, bias)
+            }
+
+        }
+      go(0, 10, in, -1)
+    }
+
+    @tailrec
+    def go(ind: Int, in: List[Char], out: List[String]): List[String] = {
```
Comment (kustosz) [inline Syntax/definition/src/main/scala/org/enso/syntax/text/Debug.scala]: This code is completely incomprehensible. Variables have no names, arguments have no names. The same is true for the method above. Special shoutout to `go(0,10,in,-1)` on line 34, what happens if I change `10` to `-4`?

File: azure-pipelines.yml [hunk 10]
```diff
@@ -44,10 +44,10 @@ jobs:
       sbt -no-colors syntax/bench
     displayName: sbt bench parser
     continueOnError: true
-  - script: |
-      sbt -no-colors interpreter/bench
-    displayName: sbt bench interpreter
-    continueOnError: true
+#  - script: |
```
Comment (kustosz) [inline azure-pipelines.yml]: Uncomment or review, make a decision! :)

File: Interpreter/src/main/java/org/enso/interpreter/node/ProgramRootNode.java (hunk 7 above)
Comment (iamrecursion) [reply Interpreter/src/main/java/org/enso/interpreter/node/ProgramRootNode.java]: I swear I had a deopt here at one point. No idea what happened.

File: azure-pipelines.yml (hunk 10 above)
Comment (iamrecursion) [reply azure-pipelines.yml]: Benchmarks are now disabled on CI.

File: Syntax/definition/src/main/scala/org/enso/syntax/text/Debug.scala (hunk 9 above)
Comment (iamrecursion) [reply Syntax/definition/src/main/scala/org/enso/syntax/text/Debug.scala]: This isn't my code, unfortunately. Haven't the foggiest how it works, I just moved it around.

**Pull request merged**